{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BA10F - Profile HMM with pseudocounts #\n",
    "\n",
    "Aligning a protein against a profile HMM\n",
    "Given a protein family, represented by Alignment, we can now return to the problem of deciding whether a newly sequenced protein, represented by Text, belongs to the family. We first form HMM(Alignment, θ) for some parameter θ. As shown in the figure below, a hidden path through HMM(Alignment, θ) corresponds to a sequence of match, insertion, and deletion states for aligning Text against Alignment.\n",
    "\n",
    "\n",
    "<br>\n",
    "<br> <img src = http://bioinformaticsalgorithms.com/images/HMM/aligning_text_against_profile.png width = 500px> \n",
    "<br>\n",
    "\n",
    "\n",
    "Figure: (Top) A path through HMM(Alignment, 0.35) for the multiple alignment from the previous section and the emitted string Text = ACAFDEAF. (Bottom) The emitted symbols correspond to aligning Text against Alignment. Specifically, the first two symbols are emitted from two match states and belong in the first two positions of the alignment. The next two symbols are emitted from an insertion state and belong in columns of their own (shown in pink). The space symbols in the seventh and eleventh columns above correspond to deletion states; these symbols are not emitted by the HMM. The space symbols in the gray columns do not correspond to any states and are passed over. The non-shaded columns form an augmented 6 × 8 seed alignment for comparison against newly sequenced proteins.\n",
    "\n",
    "To find the “best” alignment of Text against Alignment, we simply need to apply the Viterbi algorithm to find an optimal hidden path in HMM(Alignment, θ). If the product weight of this optimal hidden path exceeds a predetermined threshold, then we may conclude that Text belongs to the protein family, in which case we augment the existing seed alignment with an additional row corresponding to Text. In this way, we can recruit more and more distant family members to a seed alignment, adding these new proteins to the growing multiple alignment, and thus making the resulting profile HMM more and more suitable for analyzing the protein family of interest.\n",
    "\n",
    "￼STOP and Think: If the product weight for a new protein exceeds a threshold for more than one protein family, how would you classify this protein?\n",
    "\n",
    "Profile HMMs have finally helped us achieve our original goal of scoring different columns of a multiple alignment differently based on the frequency of symbols in each column. For example, say that the seventh column of Alignment∗ contains more occurrences of A than C, and the ninth column of Alignment∗ contains more occurrences of C than A. A hidden path passing through Match(7) would be rewarded more for emitting A than C, whereas a hidden path passing through Match(9) would be rewarded more for emitting C than A.\n",
    "\n",
    "#### The return of pseudocounts\n",
    "\n",
    "\n",
    "Returning to the figure from the previous lesson showing a table of transition probabilities, the majority of transition probabilities in the gray cells of this figure are equal to zero. (The same is true of emission probabilities.)\n",
    "\n",
    "<br>\n",
    "<img src = http://bioinformaticsalgorithms.com/images/HMM/transition_probabilities.png width = 500 px>\n",
    "<br>\n",
    "\n",
    "These zeroes may cause problems; for example, the path in the figure from aligning a string against a profile matrix, reproduced below, seems perfectly reasonable for Text = ACAFDEAF, and yet Pr(x, π) is equal to zero because the transition probability from Match(2) to Insertion(2) for this profile HMM is zero.\n",
    "\n",
    "\n",
    "As when we searched for motifs, we will introduce pseudocounts by adding a small value σ to entries in the transition matrix that correspond to edges of the HMM diagram  (i.e., only the gray elements of the table from the previous step, reproduced below). Note that white cells in this table, corresponding to forbidden transitions, are not affected by pseudocounts. The resulting matrix will then need to be normalized so that the elements in each row sum to 1.\n",
    "\n",
    "﻿Exercise Break: Compute the normalized matrix for the matrix below after adding the pseudocount σ = 0.01.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def parse_HMM(file): # with pseudocount sigma\n",
    "    lines = f.readlines()\n",
    "    # if error: check if tab or space in this input line\n",
    "    (theta, sigma) = (float(i) for i in lines[0].strip().split('\\t'))\n",
    "    alpha = lines[2].strip().split()\n",
    "    alpha = dict(zip(alpha, range(len(alpha))))\n",
    "    alignment = [line.strip() for line in lines[4:]]\n",
    "    print(\"Alignment:\")\n",
    "    for a in alignment:\n",
    "        print('\\t', a)\n",
    "    print(\"Alpha:\", alpha)\n",
    "    print(\"Theta:\", theta)\n",
    "    print(\"Sigma:\", sigma)\n",
    "    return alignment, alpha, theta, sigma\n",
    "\n",
    "### algorithm for building profile HMM from an alignment|theta\n",
    "def profileHMM(alignment, alpha, theta):\n",
    "\n",
    "    # # 'seed' indicates which columns pass threshold theta for align*\n",
    "    def get_seed_alignment(alignment, theta, alphabet):\n",
    "        \"\"\"'seed': True if gaps < theta; included in align*\n",
    "            returns seed alignment columns for HMM\"\"\"\n",
    "        k = len(alignment[0])\n",
    "        a = len(alphabet.keys())\n",
    "        print(\"Length of alignment(k) = \",k)\n",
    "        print(\"size of alphabet = \",k,'\\n')\n",
    "        freq = np.zeros(shape=(a+1, k)) \n",
    "\n",
    "        for seq in alignment:\n",
    "            for i in range(k):\n",
    "                if seq[i] == '-':\n",
    "#                     print('\\tgap', seq[i])\n",
    "                    freq[a][i] += 1\n",
    "                else:\n",
    "                    freq[alphabet[seq[i]]][i] +=1\n",
    "#                     print('\\tbase', seq[i])\n",
    "        n = len(alignment)\n",
    "        print('freq matrix:\\n', freq)\n",
    "        seed = [x/n < theta for x in freq[a]]\n",
    "        return seed\n",
    "    \n",
    "    def normalize_matrices(T,E):\n",
    "        \"\"\"Normalize Transmission and Emission frequencies to sum of row for state\"\"\"\n",
    "        for state in range(len(S)):\n",
    "            if sum(T[state]) > 0:\n",
    "                T[state] =  T[state]/sum(T[state])\n",
    "            if sum(E[state]) >0:\n",
    "                E[state] =  E[state]/sum(E[state])\n",
    "        return T, E\n",
    "    \n",
    "    def state_transition(T, prev, kind, S):\n",
    "        \"\"\" find next Match, Del, Ins, or End state in [states] \"\"\"\n",
    "        x=0\n",
    "        if S[prev][0] == 'M':\n",
    "            x=1\n",
    "        for nxt in range(prev+1+x, len(T)):\n",
    "            if S[nxt][0] == kind[0]:\n",
    "                T[prev][nxt] += 1\n",
    "#                 print('  >Transition added:',((S[prev],prev),(S[nxt], nxt)))\n",
    "                return T, nxt\n",
    "    \n",
    "    ### Walk each sequence to count transitions, emissions\n",
    "    seed = get_seed_alignment(alignment, theta, alpha)\n",
    "    print('seed:',seed)\n",
    "    n = len(alignment)\n",
    "    k = len(alignment[0])\n",
    "    S = ['S','I0']+list(c+str(n) for n in range(1,sum(seed)+1) for c in \"MDI\")+['E']\n",
    "    E = np.zeros(shape=(len(S), len(alpha.keys())))\n",
    "    T = np.zeros(shape=(len(S), len(S)))\n",
    "\n",
    "    for seq in alignment:\n",
    "        # 'state' is the hidden state col/row; i is pos in alignment\n",
    "        state = 0; i = 0\n",
    "        while i < k:\n",
    "            # seed: either match or del as hidden state\n",
    "            if seed[i]:\n",
    "                if seq[i] in alpha:\n",
    "                    T, state = state_transition(T, state, 'Match', S)\n",
    "                    E[state][alpha[seq[i]]] +=1\n",
    "                else:\n",
    "                    T, state = state_transition(T, state, 'Deletion', S)\n",
    "            # not seed: either insert or nothing\n",
    "            else:\n",
    "                # count any emissions before next seed column\n",
    "                emits = []\n",
    "                while not seed[i]:\n",
    "                    if seq[i] in alpha:\n",
    "                        emits.append(seq[i])\n",
    "                    i += 1\n",
    "                    if i == k: \n",
    "                        break  # hit end of sequence\n",
    "                i-=1                \n",
    "                if len(emits)>0:\n",
    "                    # count the transition(state, insertion)\n",
    "                    T, state = state_transition(T, state, 'Insert', S)\n",
    "                    # get all emissions(state) in insertion\n",
    "                    for symbol in emits:\n",
    "                        E[state][alpha[symbol]] += 1\n",
    "                    # count all symbols as cyclic transition t(ins_x, ins_x)\n",
    "                    if len(emits) >1:\n",
    "                        T[state][state] += len(emits)-1\n",
    "                # else do nothing, just gaps in align\n",
    "            i += 1\n",
    "        # from last state to 'End'    \n",
    "        T, state = state_transition(T, state, 'End', S)\n",
    "\n",
    "    T,E = normalize_matrices(T,E)\n",
    "    return S, T, E\n",
    "\n",
    "def pseudo_normalize(T, E):\n",
    "    \"\"\"adds pseudocounts to HMM matrices\"\"\"\n",
    "    # add pseudos and normalize E\n",
    "    for i in range(1, len(S)-1):\n",
    "        if S[i][0] != 'D':\n",
    "            E[i] += 0.01\n",
    "            E[i] = E[i]/ sum(E[i])\n",
    "    # add pseudos and normalize T\n",
    "    for i in range(len(S)-1):\n",
    "        start = 3*((i+1)//3)+1\n",
    "        end = start+3\n",
    "        for j in range(start, end):\n",
    "            if j < len(S):\n",
    "                T[i][j] += 0.01\n",
    "        T[i] = T[i]/sum(T[i])\n",
    "    return T, E\n",
    "\n",
    "def outputHMM(T,E,S,alpha):\n",
    "    \"\"\"formats the T,E matrices for Rosalind output\"\"\"\n",
    "    lists = []\n",
    "    lists.append(['']+S)\n",
    "    for i in range(len(S)):\n",
    "        # use {0:g} format to remove trailing zeros\n",
    "        lists.append([S[i]]+['{0:g}'.format(x) for x in T[i]])\n",
    "    lists.append('--------')\n",
    "    lists.append([''] + alpha)\n",
    "    for i in range(len(S)):\n",
    "        lists.append([S[i]]+['{0:g}'.format(x) for x in E[i]])\n",
    "    return lists"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Test dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment:\n",
      "\t ED-BCBDAC\n",
      "\t -D-ABBDAC\n",
      "\t ED--EBD-C\n",
      "\t -C-BCB-D-\n",
      "\t AD-BC-CA-\n",
      "\t -DDB-BA-C\n",
      "Alpha: {'A': 0, 'B': 1, 'C': 2, 'D': 3, 'E': 4}\n",
      "Theta: 0.399\n",
      "Sigma: 0.01\n",
      "Length of alignment(k) =  9\n",
      "size of alphabet =  9 \n",
      "\n",
      "freq matrix:\n",
      " [[1. 0. 0. 1. 0. 0. 1. 3. 0.]\n",
      " [0. 0. 0. 4. 1. 5. 0. 0. 0.]\n",
      " [0. 1. 0. 0. 3. 0. 1. 0. 4.]\n",
      " [0. 5. 1. 0. 0. 0. 3. 1. 0.]\n",
      " [2. 0. 0. 0. 1. 0. 0. 0. 0.]\n",
      " [3. 0. 5. 1. 1. 1. 1. 2. 2.]]\n",
      "seed: [False, True, False, True, True, True, True, True, True]\n"
     ]
    }
   ],
   "source": [
    "### Extra test data 1:\n",
    "\n",
    "\n",
    "with open('/Users/jasonmoggridge/Dropbox/Rosalind/Coursera_textbook_track/Course6/data/10f.extra.txt') as f:\n",
    "    alignment, alpha, theta, sigma = parse_HMM(f)\n",
    "\n",
    "S, T, E = profileHMM(alignment, alpha, theta)\n",
    "alpha = list(alpha.keys())\n",
    "T, E = pseudo_normalize(T, E)\n",
    "T = np.around(T,3)\n",
    "E = np.around(E,3)\n",
    "\n",
    "lists =outputHMM(T,E,S, alpha)\n",
    "with open('/Users/jasonmoggridge/Dropbox/Rosalind/Coursera_textbook_track/Course6/data/10f.extra.out.txt','w') as file:\n",
    "    for line in lists[:-1]:\n",
    "        if line[0] != '-':\n",
    "            file.write('\\t'.join(line))\n",
    "        else:\n",
    "            file.write(line)\n",
    "        file.write('\\n')\n",
    "    file.write('\\t'.join(lists[-1]))\n",
    "del(line, lists)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment:\n",
      "\t EBA\n",
      "\t E--\n",
      "\t -BA\n",
      "\t EBA\n",
      "\t EBA\n",
      "\t EBE\n",
      "\t ---\n",
      "\t ABA\n",
      "Alpha: {'A': 0, 'B': 1, 'C': 2, 'D': 3, 'E': 4}\n",
      "Theta: 0.274\n",
      "Sigma: 0.01\n",
      "Length of alignment(k) =  3\n",
      "size of alphabet =  3 \n",
      "\n",
      "freq matrix:\n",
      " [[1. 0. 5.]\n",
      " [0. 6. 0.]\n",
      " [0. 0. 0.]\n",
      " [0. 0. 0.]\n",
      " [5. 0. 1.]\n",
      " [2. 2. 2.]]\n",
      "seed: [True, True, True]\n"
     ]
    }
   ],
   "source": [
    "### Stepik challenge data:\n",
    "\n",
    "## 'theta sigma' line was space sep (changed parser)\n",
    "\n",
    "with open('/Users/jasonmoggridge/Dropbox/Rosalind/Coursera_textbook_track/Course6/data/dataset_26259_5.txt') as f:\n",
    "    alignment, alpha, theta, sigma = parse_HMM(f)\n",
    "\n",
    "S, T, E = profileHMM(alignment, alpha, theta)\n",
    "alpha = list(alpha.keys())\n",
    "T, E = pseudo_normalize(T, E)\n",
    "T = np.around(T,3)\n",
    "E = np.around(E,3)\n",
    "\n",
    "lists =outputHMM(T,E,S, alpha)\n",
    "with open('/Users/jasonmoggridge/Dropbox/Rosalind/Coursera_textbook_track/Course6/data/BA10F.dataset_26259_5.out.txt','w') as file:\n",
    "    for line in lists[:-1]:\n",
    "        if line[0] != '-':\n",
    "            file.write('\\t'.join(line))\n",
    "        else:\n",
    "            file.write(line)\n",
    "        file.write('\\n')\n",
    "    file.write('\\t'.join(lists[-1]))\n",
    "del(line, lists)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment:\n",
      "\t A-BEBCEEA\n",
      "\t AA-DBEEAC\n",
      "\t AAADB-EDC\n",
      "\t BABDBEC-C\n",
      "\t ABBDBEEAC\n",
      "\t AABDBEEEC\n",
      "\t AABCBEEAC\n",
      "\t AABDBEE-C\n",
      "Alpha: {'A': 0, 'B': 1, 'C': 2, 'D': 3, 'E': 4}\n",
      "Theta: 0.223\n",
      "Sigma: 0.01\n",
      "Length of alignment(k) =  9\n",
      "size of alphabet =  9 \n",
      "\n",
      "freq matrix:\n",
      " [[7. 6. 1. 0. 0. 0. 0. 3. 1.]\n",
      " [1. 1. 6. 0. 8. 0. 0. 0. 0.]\n",
      " [0. 0. 0. 1. 0. 1. 1. 0. 7.]\n",
      " [0. 0. 0. 6. 0. 0. 0. 1. 0.]\n",
      " [0. 0. 0. 1. 0. 6. 7. 2. 0.]\n",
      " [0. 1. 1. 0. 0. 1. 0. 2. 0.]]\n",
      "seed: [True, True, True, True, True, True, True, False, True]\n"
     ]
    }
   ],
   "source": [
    "### Rosalind challenge data:\n",
    "\n",
    "## 'theta sigma' line was space sep (changed parser)\n",
    "\n",
    "with open('/Users/jasonmoggridge/Dropbox/Rosalind/Coursera_textbook_track/Course6/data/rosalind_ba10f.txt') as f:\n",
    "    alignment, alpha, theta, sigma = parse_HMM(f)\n",
    "\n",
    "S, T, E = profileHMM(alignment, alpha, theta)\n",
    "alpha = list(alpha.keys())\n",
    "T, E = pseudo_normalize(T, E)\n",
    "T = np.around(T,3)\n",
    "E = np.around(E,3)\n",
    "\n",
    "lists =outputHMM(T,E,S, alpha)\n",
    "with open('/Users/jasonmoggridge/Dropbox/Rosalind/Coursera_textbook_track/Course6/data/BA10F.rosalind_ba10f.out.txt','w') as file:\n",
    "    for line in lists[:-1]:\n",
    "        if line[0] != '-':\n",
    "            file.write('\\t'.join(line))\n",
    "        else:\n",
    "            file.write(line)\n",
    "        file.write('\\n')\n",
    "    file.write('\\t'.join(lists[-1]))\n",
    "del(line, lists)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
